### PR TITLE
[WIP] Make the NSA fast-path CUDA graphable

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
## Motivation

Follow up for #23351 to bring the fast-path into the CUDA graph which yields an extra 4-8% improvement on the 1k/1k pareto.

**Note:**

_This PR has a bug in the current form:_

> _skip\_logits\_computation = x\_meta.shape[0] <= self.index\_topk_

_This is not correct for chunked prefill case. WIP for a potential work-around._

## Modifications

- Made the K-cache store CUDA-graphable by making it padding aware, added tests.
- Added a pure PyTorch implementation of the Top-K transform op. This is easier to do than to add support for CUDA-graphs in the backend top-k transform op.
- Some surgical changes / custom ops to make Dynamo happy.
- Finally, added some infra for staging static buffers for some metadata tensors required in the fast path.

AI used for code generation: Claude Opus 4.7

## Accuracy Tests

Ran GPQA eval:

```
Repeat: 8, mean: 0.851
Scores: ['0.823', '0.874', '0.854', '0.843', '0.843', '0.874', '0.859', '0.838']
```

## Speed Tests and Profiling

image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

cc @nvjullin @mattteochen @nvpohanh

### CODE REVIEW

File: python/sglang/srt/layers/attention/nsa/metadata_staging.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    # (max_num_tokens, index_topk) int32 — scratch output for
+    # `topk_construct_k_only_pcg`. Owned by the runner (mark_static_address);
+    # not graph-pool managed, so its address is stable across captures and
+    # replays.
+    topk_result: torch.Tensor
+
+    # `TopkTransformMethod` enum captured on the first `stage_*` call; pinned
+    # thereafter (drift across captures triggers an assertion). Read inside
+    # the compiled region as a Dynamo-specialized Python int.
+    topk_transform_method: Optional[Any] = None
```
Comment (tirthasheshpatel) [inline python/sglang/srt/layers/attention/nsa/metadata_staging.py:56]: I can remove this. Dynamo guards on it anyways and will re-compile if it changes. I left it since it'd better to error out than recompile.

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 2]
```diff
@@ -76,6 +80,126 @@
...
+        Only PAGED is supported under PCG: RAGGED requires the
+        flashmla_sparse prefill backend (`nsa_backend.py:get_topk_transform_method`),
+        which is not in scope here. Padded rows beyond `num_tokens` have
+        `seqlens_expanded[i] == 0` (zeroed by `_stage_1d`), so the validity
+        mask collapses them to all -1.
+
+        `topk_result` is a stable-address scratch buffer owned by the runner
+        (mark_static_address). Writing into it bypasses the CUDA-graph pool's
+        per-capture lifetime tracking, which would otherwise reuse the slab
+        before FMHA reads topk_indices under TP>1 causing IMA.
```
Comment (tirthasheshpatel) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py:179]: Using torch.full to create the result tensor inside the CUDA graph causes an IMA. It might be a bug in the CUDA caching allocator.

For now, workaround this by making it a static buffer outside the graph.

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 3]
```diff
@@ -748,8 +867,13 @@ def _forward_cuda_k_only(
         # Fast path: only compute and store k cache, skip all q and weights ops
         key = self._get_k_bf16(x, positions, enable_dual_stream)
 
-        if not forward_batch.out_cache_loc.is_contiguous():
-            forward_batch.out_cache_loc = forward_batch.out_cache_loc.contiguous()
+        if not is_in_piecewise_cuda_graph():
+            # Under PCG, out_cache_loc is the runner's pre-allocated contiguous
+            # buffer (slicing it yields a contiguous view); the check itself is
+            # Dynamo-hostile (is_contiguous is a data-dependent tensor call).
```
Comment (tirthasheshpatel) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py:873]: Will remove this, the comment is not correct (`is_contiguous` is not data-dependent). Missed it when reviewing the code generated by claude.

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 4]
```diff
@@ -762,7 +886,27 @@ def _forward_cuda_k_only(
...
+            ctx = get_forward_context()
+            bufs = ctx.nsa_metadata_buffers
+            assert (
+                bufs is not None
+            ), "PCG inlined k-only topk requires staged NSA metadata buffers"
+            method = bufs.topk_transform_method
+            assert method is not None, (
+                "PCG inlined k-only topk requires topk_transform_method to "
+                "have been pinned by stage_prefill_nsa_metadata_buffers"
+            )
```
Comment (tirthasheshpatel) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py:900]: Will remove the assertion, it always holds true.

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 5]
```diff
@@ -1118,15 +1289,15 @@ def forward_cuda(
...
-        # We can only skip the logits computation if cuda graph is not involved
+        # Determine if should skip topk based on sequence length.
         skip_logits_computation = False
         if forward_batch.forward_mode.is_extend_without_speculative():
-            if forward_batch.seq_lens_cpu is not None:
+            if is_in_piecewise_cuda_graph():
+                # This is a heuristic. At compile-time if the input sequence
+                # is <= index_topk number of tokens, we can trivially skip
+                # logits computation.
+                skip_logits_computation = x_meta.shape[0] <= self.index_topk
```
Comment (tirthasheshpatel) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py:1299]: This is the reason for the perf improvement. It's a pretty naive heuristic but powerful enough to yield the improvement

**Review by tirthasheshpatel — COMMENTED** [review]
This PR has a bug in the current form:

> skip_logits_computation = x_meta.shape[0] <= self.index_topk

This is not correct for chunked prefill case. WIP for a potential work-around.

**Comment by tirthasheshpatel** [comment]
Closed this PR since we decided to not go down this path. To make this patch usable, we would need to capture two CUDA graphs: one for the fast path and one for the slow path. @nvjullin pointed out the following limitations of this approach:

- The compilation time, PCG already takes quite some time on server start
- The memory consumption
- Combinatorial explosion: every cudagraph-incompatible branch multiplies the number of cudagraphs needed, and it'll become infeasible very fast
- There's also the problem of having the dispatch logic hoisted out of the model into the cudagraph runner code, which is architecturally questionable.

This PR helped determine that simply launching fewer kernels on the fast path is faster than CUDA graphing the slow path. @mattteochen's PR https://github.com/sgl-project/sglang/pull/25642 uses the eager fast-path implementation and supersedes this work.

@IvanYashchuk
